pkg/gatewayserver/io/basicstationlns/basicstationlns_test.go

```diff
@@ -0,0 +1,1030 @@
+				t.Fatalf("Failed to write message: %v", err)
+			}
+
+			go func() {
+				_, _, err := conn.ReadMessage()
+				if err == nil {
+					t.Fatalf("Expected connection closure with error but received none")
+				}
+			}()
+		})
+	}
+
+	// Valid
+	for i, tc := range []struct {
+		EndPointEUI string
+		EUI         types.EUI64
+		Query       interface{}
+	}{
+		{
+			EndPointEUI: "1111111111111111",
+			EUI:         types.EUI64{0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11},
+			Query: messages.DiscoverQuery{
+				EUI: basicstation.EUI{
+					Prefix: "router",
+					EUI64:  types.EUI64{0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11},
+				},
+			},
+		},
+	} {
+		t.Run(fmt.Sprintf("ValidQuery/%d", i), func(t *testing.T) {
+			a := assertions.New(t)
+			conn, _, err := websocket.DefaultDialer.Dial(discoveryEndPoint, nil)
+			if !a.So(err, should.BeNil) {
+				t.Fatalf("Connection failed: %v", err)
+			}
+			defer conn.Close()
+			req, err := json.Marshal(tc.Query)
+			if err != nil {
+				panic(err)
+			}
+			if err := conn.WriteMessage(websocket.TextMessage, req); err != nil {
+				t.Fatalf("Failed to write message: %v", err)
+			}
+
+			resCh := make(chan []byte)
+			go func() {
+				_, data, err := conn.ReadMessage()
+				if err != nil {
+					close(resCh)
+					if err == websocket.ErrBadHandshake {
+						return
+					} else {
+						t.Fatalf("Failed to read message: %v", err)
+					}
+				}
+				resCh <- data
+			}()
+			select {
+			case res := <-resCh:
+				var response messages.DiscoverResponse
+				if err := json.Unmarshal(res, &response); err != nil {
+					t.Fatalf("Failed to unmarshal response `%s`: %v", string(res), err)
+				}
+				a.So(response, should.Resemble, messages.DiscoverResponse{
+					EUI: basicstation.EUI{Prefix: "router", EUI64: tc.EUI},
+					Muxs: basicstation.EUI{
+						Prefix: "muxs",
+					},
+					URI: connectionRootEndPoint + "eui-" + tc.EndPointEUI,
+				})
+			case <-time.After(timeout):
+				t.Fatalf("Read message timeout")
+			}
+			conn.Close()
+		})
+	}
+}
+
+func TestVersion(t *testing.T) {
+	a := assertions.New(t)
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+
+	c := component.MustNew(test.GetLogger(t), &component.Config{
+		ServiceBase: config.ServiceBase{
+			HTTP: config.HTTP{
+				Listen: ":8100",
+			},
+		},
+	})
+	gs := mock.NewServer()
+	srv := New(ctx, gs)
+	c.RegisterWeb(srv)
+	test.Must(nil, c.Start())
+	defer c.Close()
+	gs.RegisterGateway(ctx, registeredGatewayID, &registeredGateway)
+
+	conn, _, err := websocket.DefaultDialer.Dial(testTrafficEndPoint, nil)
+	if !a.So(err, should.BeNil) {
+		t.Fatalf("Connection failed: %v", err)
+	}
+	defer conn.Close()
+
+	for _, tc := range []struct {
+		Name                 string
+		VersionQuery         interface{}
+		ExpectedRouterConfig interface{}
+	}{
+		// These tests will fail without unmarshalling of Sx1301_conf https://github.com/TheThingsNetwork/lorawan-stack/issues/808
```

Comment:
So we can't test the router config without this issue? Seems quite important to me

Reply:
router config is tested and verified on the TTIG. The only reason why un-marshaling of SX1301_conf is not implemented is due to performance reasons (of the un-marshaler). I don't think we need to block this PR for that.

Reply:
Performance reasons?

Reply:
from @htdvisser :
> 1. Because we didn't actually have a need to unmarshal it
> 2. Because the stdlib encoding/json package would be inefficient at unmarshaling it

This is why the linked issue was created. I can simply use an "inefficient" marshaller for now but only if it's really needed.